feat(cache): implement sprite caching

CommanderStorm opened the pull request:
This PR adds sprite caching support to Martin.

Resolves #1342

CommanderStorm committed:
renaming

CommanderStorm committed:
draft

CommanderStorm committed:
tmp

CommanderStorm commented on martin/src/config/file/main.rs:
```diff
@@ -182,44 +196,95 @@ impl Config {
+            let font_cache_size_mb = if let FileConfigEnum::Config(cfg) = self.fonts {
+                cfg.custom.cache_size_mb.unwrap_or(cache_size_mb / 8) // Default: 12.5% for fonts
+            } else {
+                cache_size_mb / 8 // Default: 12.5% for fonts
+            };
+
+            CacheConfig {
+                #[cfg(any(
+                    feature = "postgres",
+                    feature = "pmtiles",
+                    feature = "mbtiles",
+                    feature = "unstable-cog"
+                ))]
+                tile_cache_size_mb: self.tile_cache_size_mb.unwrap_or(cache_size_mb / 2), // Default: 50% for tiles
+                #[cfg(feature = "pmtiles")]
+                pmtiles_cache_size_mb,
+                #[cfg(feature = "sprites")]
+                sprite_cache_size_mb,
+                #[cfg(feature = "fonts")]
+                font_cache_size_mb,
+            }
+        } else {
+            // TODO: the defaults could be smarter. If I don't have pmtiles sources, don't reserve cache for it
+            CacheConfig {
+                #[cfg(any(
+                    feature = "postgres",
+                    feature = "pmtiles",
+                    feature = "mbtiles",
+                    feature = "unstable-cog"
+                ))]
+                tile_cache_size_mb: 256,
+                #[cfg(feature = "pmtiles")]
+                pmtiles_cache_size_mb: 128,
+                #[cfg(feature = "sprites")]
+                sprite_cache_size_mb: 64,
+                #[cfg(feature = "fonts")]
+                font_cache_size_mb: 64,
+            }
+        }
+    }
```
@nyurik Could you review the basic concept of this work?

I split the cache into multiple parts.
The PR is not yet in a reviewable state, but if the basic concept of this work is something we can sync on would still be quite nice to know 😉

nyurik replied:
seems reasonable.  We could of course have simpler division of tiles vs everything else, but giving a bit more control is probably ok.  Reducing size is probably not worth it if we do not pre-allocate memory. The only optimization here might be due async cache purging - if we have a dedicated process to purge cache (doubt it), it might be worth it to not even start that cache sub-system. Apart from that, probably not worth it.

CommanderStorm committed:
more fixes

CommanderStorm committed:
tmp

CommanderStorm committed:
add a _tiles feature

CommanderStorm committed:
remove _tiles from cargo hack

CommanderStorm committed:
Merge branch '_tiles-feature' into no-main-cache

CommanderStorm committed:
formatting

CommanderStorm committed:
fix clippy lints

CommanderStorm committed:
Merge branch 'main' into no-main-cache

CommanderStorm committed:
docs(pg): clarify the `properties` config file entry (#2300)

CommanderStorm committed:
docs: clearer pg function documentation (#2301)

I got bitten a bit by there not existing a lot of documentation on how
postgres sources have to be written and why.

This PR tries to improve this

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

CommanderStorm committed:
tests(pg) Update PostGIS image to version 18-3.6 in CI and docs (#2259)

This update ensures that we also work with pg 18 / 3.6

CommanderStorm committed:
chore: move `MainCache` to be a `TileCache` (#2297)

This PR is one of these PRs, which by itsself does not make sense.

This makes sense to look at if the following PR is accepted, as it makes
the diff smaller:

- #2295

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

CommanderStorm committed:
Merge branch 'main' into no-main-cache

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

CommanderStorm committed:
fix fmt

CommanderStorm committed:
Merge branch 'main' into no-main-cache

CommanderStorm committed:
self-review

CommanderStorm committed:
fix serialisation issue

CommanderStorm committed:
apply suggestions from code review

CommanderStorm committed:
Merge branch 'main' into no-main-cache

CommanderStorm commented on martin-core/src/resources/sprites/cache.rs:
```diff
@@ -0,0 +1,131 @@
+use moka::future::Cache;
+
+/// Sprite cache for storing generated sprite sheets.
+
```
```suggestion
```

CommanderStorm committed:
Apply suggestion from @CommanderStorm

CommanderStorm commented on martin/src/config/file/resources/fonts.rs:
```diff
@@ -1,5 +1,3 @@
-use std::collections::BTreeMap;
-
 use martin_core::fonts::FontSources;
```
```suggestion
use std::collections::BTreeMap;
use martin_core::fonts::FontSources;
```

CommanderStorm committed:
Update martin/src/config/file/resources/fonts.rs

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

nyurik commented on martin-core/src/resources/sprites/cache.rs:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    /// Inserts a sprite sheet into the cache.
+    pub async fn insert(&self, ids: &str, as_sdf: bool, as_json: bool, bytes: Bytes) {
```
if you need an owned param, you should ask for it. I.e. - rule of thumb - if you have to do `ids.to_string()`, just ask for `String` as the param

nyurik commented on martin/src/config/file/cache.rs:
```diff
@@ -18,8 +21,11 @@ impl CacheConfig {
     pub fn create_tile_cache(&self) -> Option<martin_core::tiles::TileCache> {
         if self.tile_cache_size_mb > 0 {
+            log::info!(
+                "Initializing tile cache with maximum size {} MiB",
```
can we please just use `MB`? Makes life soo much more readable :)

CommanderStorm replied:
Read:
- change to base `1000` too or
- stick with base `1024` and incorrectly/windows-style mark this as `MB`

CommanderStorm replied:
Our frontend also uses `MiB`, so better change there too

CommanderStorm replied:
mbtiles also uses `KiB` => via `size_format::SizeFormatterBinary`

nyurik replied:
sigh, just remember that our config values are `_mb` -- so we want consistency.  Also, very few ppl use `MiB` - this is one of those "technically accurate, but who cares" type of things :)

CommanderStorm replied:
The difference is a 5% larger or smaller cache.

I adjusted it to be option 1 => change to base 1000 too.
That seems to be the most correct thing to do

nyurik reviewed (approved):
minor nits

CommanderStorm committed:
move the cloning one level up to make it more apparent

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

CommanderStorm committed:
fix MiB vs MB inconsistency

CommanderStorm committed:
fix typo in fronted tests

CommanderStorm committed:
make the "Compare test output results" simpler to debug

CommanderStorm committed:
update the snapshots
